pickle/src/com/treode/pickle/PicklerRegistry.scala

```diff
@@ -14,100 +14,269 @@
+ * 
+ * @param  p The Pickler to add to the HashMap
+ * @param random A tuple of the Id and the reader method to use for the Pickler
+ *               to register
+ * @return Long Returns the Id of the pickler that was added or the old Opener
+ *         if there was already a Pickler associated with that Id
+ */
+ def open [P] (p: Pickler [P]) (random: => (Long, P => T)): Long = {
+  var r = random
+  var u1 = Opener (p, r._1, r._2)
+  var u0 = openers.putIfAbsent (r._1, u1)
+  while (u0 != null) {
+    r = random
+    u1 = Opener (p, r._1, r._2)
+    u0 = openers.putIfAbsent (r._1, u1)
   }
+  r._1
+}
+
+/**
+ *  Remove Pickle with the id ``tag``
+ *  
+ * @param tag the id of the unpickler to remove
+ * @return Unit Returns nothing
+ */
+ def unregister (tag: Long): Unit =
+ openers.remove (tag)
+
+/**
+ *  Unpickles and leaves readPos wherever after getting the opener
+ *  for the pickler with the id ``id`` and unserializing the InputBuffer
+ *  ``buf`` - it doesn't check for any errors and leaves the read position
+ *  wherever it lands
+ *  
+ * @param id the id of the unpickler to use
+ * @param buf the InputBuffer to use
+ */
+ def unpickle(id: Long, buf:InputBuffer):T = {
+  val u = openers.get (id)
+  if (u == null)
+  return default (id)
+  val v = u.read (buf)
+  v
+}
+
+/**
+ *  This unpickle method reads a long id from the buffer to find
+ *  which Pickler to use and then unpickles the rest of the buffer
+ *  using the above function - passing the rest of the buffer as the
+ *  InputBuffer to read from.
+ * @param buf the InputBuffer to use
+ * @return T The unpickled, read object
+ */
+ def unpickle(buf:InputBuffer):T = 
+ unpickle (buf.readLong(), buf)
 
-  def open [P] (p: Pickler [P]) (random: (Long, P => T)): Long = {
-    var r = random
-    var u1 = Opener (p, r._1, r._2)
-    var u0 = openers.putIfAbsent (r._1, u1)
-    while (u0 != null) {
-      r = random
-      u1 = Opener (p, r._1, r._2)
-      u0 = openers.putIfAbsent (r._1, u1)
-    }
-    r._1
+/**
+* This unpickle method checks that len is consumed. 
+* If its not, we then move the 
+* read header in order to discard the bytes appropriate
+* number of bytes given by ``len``.
+* Sets buf.readPos as though len had been consumed
+* and throws an exception. We assume len is the length
+* of the buffer and you want to consume all of the buffer
+ *  
+ * @param buf the InputBuffer to use
+ * @param id the id of the unpickler to use
+ * @return T The unpickled, read object
+ */
+ def unpickle (id: Long, buf: InputBuffer, len: Int): T = {
+  val end = buf.readPos + len
+  val u = openers.get (id)
+  if (u == null) {
+    buf.readPos = end
+    return default (id)
+  }
+  val v = u.read (buf)
+  if (buf.readPos != end) {
+    buf.readPos = end
+    throw new FrameBoundsException
   }
+  v
+}
 
-  def unregister (tag: Long): Unit =
-    openers.remove (tag)
-
-  def unpickle (id: Long, buf: PagedBuffer, len: Int): T = {
-    val end = buf.readPos + len
```

Comment:
I think there's a way to factor the common code between this method and the unpickle (InputBuffer) above.